### PR TITLE
ZK-3543: Can't wire <apply> with its ID

### PR DESCRIPTION
**Pull request by DevChu**

### CODE REVIEW

File: zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java [hunk 1]
```diff
@@ -291,51 +305,118 @@ private ComponentMatchCtx buildNextCtx() {
...
+			} else {
+				break; //found
+			}
+		}
+		return child;
+	}
+
+	private Component getNextUntrackedDistributedChild(List<Component> distributedChildren) {
+		for (int i = 0, size = distributedChildren.size(); i < size; i++) {
+			Component comp = distributedChildren.get(i);
```
Comment (jumperchen) [inline zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java:334]: should not use get(i) here, please check how the implementation of getDistributedChildren() is.

File: zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java [hunk 2]
```diff
@@ -291,51 +305,118 @@ private ComponentMatchCtx buildNextCtx() {
...
+			if (!_trackedUuid.contains(comp.getUuid())) {
+				return comp;
+			}
+		}
+		return null;
+	}
+
+	private ShadowElement getNextUntrackedShadowRoot(List<ShadowElement> shadowRoots) {
+		for (int i = 0, size = shadowRoots.size(); i < size; i++) {
+			ShadowElement shadowElement = shadowRoots.get(i);
```
Comment (jumperchen) [inline zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java:344]: @DevChu  should not use get(i) here, as same as the above comment

File: zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java [hunk 3]
```diff
@@ -291,51 +305,118 @@ private ComponentMatchCtx buildNextCtx() {
...
+			if (componentMatchCtx != null) {
+				return componentMatchCtx;
+			}
+		}
+		_currCtx = _currCtx.getParent();
+		if (_currCtx == null) {
+			return null;
+		} else {
+			Component currComp = _currCtx.getComponent();
+			if (getNextUntrackedChild(currComp) == null && (!_currCtx.isShadowHost() || _currCtx.isShadowHost() && getNextUntrackedShadowRoot(((ComponentCtrl) currComp).getShadowRoots()) == null) && currComp == (_posOffset > 0 ? _offsetRoot : _root)) {
```
Comment (jumperchen) [inline zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java:382]: > (!_currCtx.isShadowHost() || _currCtx.isShadowHost() &&

strange logic here.

**Review by jumperchen — APPROVED** [review]

File: zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java [hunk 4]
```diff
@@ -291,51 +311,125 @@ private ComponentMatchCtx buildNextCtx() {
...
-			if (host instanceof ComponentCtrl) {
-				List<ShadowElement> seRoots = ((ComponentCtrl) host).getShadowRoots();
-				if (seRoots != null && seRoots.size() > 1) { //if equal to 1, then it is the current comp itelf
-					int index = seRoots.indexOf(comp) + 1;
-					if (index < seRoots.size()) { //
-						return seRoots.get(index);
-					}
+		boolean isShadow = comp instanceof ShadowElement;
+		Component child = getNextUntrackedChild(comp);
+		if (!isShadow) {
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/select/impl/ComponentIterator.java:356]: refine

**Pull request merged**